core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java

```diff
@@ -157,12 +165,13 @@ private Object convertJsonNode(JsonNode val)
       Map<String, Object> newMap = new LinkedHashMap<>();
       for (Iterator<Map.Entry<String, JsonNode>> it = val.fields(); it.hasNext(); ) {
         Map.Entry<String, JsonNode> entry = it.next();
-        newMap.put(entry.getKey(), valueConversionFunction(entry.getValue()));
+        newMap.put(entry.getKey(), finalizeConversionForMap(entry.getValue()));
       }
       return newMap;
     }
 
-    return val;
+    // turn anything else into a string value so that we don't leak JsonNode into rows
+    return val.asText();
```

Comment:
Javadoc for `asText` says:

> Method that will return a valid String representation of the container value, if the node is a value node (method isValueNode returns true), otherwise empty String.

Since we're checking for boolean, number, null, and string already, I don't think there are any cases here where `val.asText()` will return anything other than an empty string. Perhaps we should return `null` here, instead, then?

Reply:
Yeah, I agree we are handling all of the cases of value types, and arrays and objects are handled, so returning `null` seems reasonable for the special Jackson node types.